src/table-index.ts

```diff
@@ -0,0 +1,149 @@
+import { Column, ColumnDefinition, ColumnDefinitionsToColumns } from './column';
+import { DefaultExpression, Expression } from './expression';
+import { toSnakeCase, wrapQuotes } from './naming';
+import { GroupToken, SeparatorToken, StringToken, Token } from './tokens';
+
+export type IndexDefinitionsToIndexes<
+  TableNameT extends string,
+  IndexDefinitionsT extends { [column: string]: IndexDefinition<boolean, boolean> },
+> = {
+  [IndexName in keyof IndexDefinitionsT]: IndexName extends string
+    ? IndexDefinitionsT[IndexName] extends IndexDefinition<infer IsPrimaryKey, infer IsUniqueKey>
+      ? Index<IndexName, TableNameT, IsPrimaryKey, IsUniqueKey>
+      : never
+    : never;
+};
+
+export interface IndexDefinitionFormat {
+  type: string;
+  expressions: Expression<any, boolean, string>[];
+  isUniqueKey: boolean;
+  isPrimaryKey: boolean;
+  include: Column<string, string, any, boolean, boolean, any>[];
+  where: DefaultExpression<boolean> | null;
+}
+
+export interface IndexDefinition<
+  IsPrimaryKey extends boolean = false,
+  IsUniqueKey extends boolean = false,
+> {
+  getDefinition(): IndexDefinitionFormat;
+  where(condition: DefaultExpression<boolean>): IndexDefinition<IsPrimaryKey, IsUniqueKey>;
+  primaryKey(): IndexDefinition<true, true>;
+  unique(): IndexDefinition<IsPrimaryKey, true>;
+  include(
+    ...columns: Column<string, string, any, boolean, boolean, any>[]
+  ): IndexDefinition<IsPrimaryKey, IsUniqueKey>;
+}
+
+export class Index<
+  Name extends string,
+  TableName extends string,
+  IsPrimaryKey extends boolean,
+  IsUnique extends boolean,
+> {
+  private _indexBrand: any;
+
+  /** @internal */
+  getName() {
+    return this.indexName;
+  }
+
+  constructor(
+    private readonly indexName: Name,
+    private readonly tableName: TableName,
+    private readonly indexType: string,
+    private readonly isPrimaryKey: IsPrimaryKey,
+    private readonly isUniqueKey: IsUnique,
+    private readonly expressions: Expression<any, boolean, string>[],
+    private readonly include: Column<string, TableName, any, boolean, boolean, any>[],
+    private readonly where: DefaultExpression<boolean> | null,
+  ) {}
+
+  /** @internal */
+  toTokens(): Token[] {
```

Comment:
This function seems kind of sloppy, but keeping it as tokens gives us the advantage of being able to use `makeQueryState()` to keep track of columns referenced in the index if we want to implement a query-efficiency heuristic later on.